resource-managers/kubernetes/core/src/main/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleExecutorComponents.scala

```diff
@@ -0,0 +1,102 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle
+
+import java.io.File
+import java.util.Optional
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.{SparkConf, SparkEnv}
+import org.apache.spark.internal.Logging
+import org.apache.spark.shuffle.api.{ShuffleExecutorComponents, ShuffleMapOutputWriter, SingleSpillShuffleMapOutputWriter}
+import org.apache.spark.shuffle.sort.io.LocalDiskShuffleExecutorComponents
+import org.apache.spark.storage.{BlockId, BlockManager, StorageLevel, UnrecognizedBlockId}
+import org.apache.spark.util.Utils
+
+class KubernetesLocalDiskShuffleExecutorComponents(sparkConf: SparkConf)
+  extends ShuffleExecutorComponents with Logging {
+
+  private val delegate = new LocalDiskShuffleExecutorComponents(sparkConf)
+  private var blockManager: BlockManager = _
+
+  override def initializeExecutor(
+      appId: String, execId: String, extraConfigs: java.util.Map[String, String]): Unit = {
+    delegate.initializeExecutor(appId, execId, extraConfigs)
+    blockManager = SparkEnv.get.blockManager
+    if (sparkConf.getBoolean("spark.kubernetes.driver.reusePersistentVolumeClaim", false)) {
+      // Turn off the deletion of the shuffle data in order to reuse
+      blockManager.diskBlockManager.deleteFilesOnStop = false
+      Utils.tryLogNonFatalError {
+        KubernetesLocalDiskShuffleExecutorComponents.recoverDiskStore(sparkConf, blockManager)
+      }
+    }
+  }
+
+  override def createMapOutputWriter(shuffleId: Int, mapTaskId: Long, numPartitions: Int)
+    : ShuffleMapOutputWriter = {
+    delegate.createMapOutputWriter(shuffleId, mapTaskId, numPartitions)
+  }
+
+  override def createSingleFileMapOutputWriter(shuffleId: Int, mapId: Long)
+    : Optional[SingleSpillShuffleMapOutputWriter] = {
+    delegate.createSingleFileMapOutputWriter(shuffleId, mapId)
+  }
+}
+
+object KubernetesLocalDiskShuffleExecutorComponents extends Logging {
+  /**
+   * This tries to recover shuffle data of dead executors' local dirs if exists.
+   * Since the executors are already dead, we cannot use `getHostLocalDirs`.
+   * This is enabled only when spark.kubernetes.driver.reusePersistentVolumeClaim is true.
+   */
+  def recoverDiskStore(conf: SparkConf, bm: BlockManager): Unit = {
+    // Find All files
+    val files = Utils.getConfiguredLocalDirs(conf)
+      .filter(_ != null)
+      .map(s => new File(new File(new File(s).getParent).getParent))
+      .flatMap { dir =>
+        val oldDirs = dir.listFiles().filter { f =>
+          f.isDirectory && f.getName.startsWith("spark-")
+        }
+        val files = oldDirs
+          .flatMap(_.listFiles).filter(_.isDirectory) // executor-xxx
+          .flatMap(_.listFiles).filter(_.isDirectory) // blockmgr-xxx
+          .flatMap(_.listFiles).filter(_.isDirectory) // 00
```

Comment:
Do we need to check the directory names too?

Reply:
Yes, it's a protected approach which excludes some random files in the intermediate level.

Reply:
Oh, will you add check directory name (executor-xxx, blockmgr-xxx, etc) later?

Reply:
Sure, we can check the pattern. I will do later~